File: backend-cluster/ledger/src/foundation/managed-prices/managed-price-cache.ts
import type { ManagedPricesConfig } from "@/config";
import { sliceSha256 } from "@/foundation/rustledger/source-slice";
import { CACHE_KEYS, TTL, type CacheHelper } from "@/shared/cache";
import { lock } from "@/shared/lock";
import { logger } from "@/shared/logger";
import {
  fetchManagedPriceFeed,
  validateManagedPriceText,
  type ManagedPriceFeedSummary,
  type ManagedPriceValidation,
} from "./managed-price-feed";

const log = logger.child({ module: "managed-prices" });
const FEED_LOCK_PREFIX = "managed-price:";

/**
 * Feed cache for managed price includes (ADR 015 section 5): an immutable
 * blob per validated revision plus a small mutable head per URL. Refresh is
 * driven by `head.nextRefreshAt` (time stamps), never by key expiry, and a
 * failed or invalid refresh only records the error while the previous
 * revision keeps serving. Concurrent loads for one URL share a single fetch
 * through the process lock, across every ledger on the node.
 */

interface PriceFeedBlob {
  url: string;
  revision: string;
  etag: string | null;
  text: string;
  /** Epoch ms of the fetch that produced these bytes. */
  fetchedAt: number;
  feed: ManagedPriceFeedSummary;
}

export interface PriceFeedHead {
  /** The revision currently serving; null until one validates. */
  revision: string | null;
  /** Epoch ms after which the next load re-fetches. */
  nextRefreshAt: number;
  /** Why the last refresh failed, cleared by the next success. */
  lastError: string | null;
}

export interface ManagedPriceFeedDeps {
  cache: CacheHelper;
  config: ManagedPricesConfig;
  /** Injectable clock (epoch ms) so tests control refresh windows. */
  now?: () => number;
  fetchImpl?: typeof fetch;
}

export interface ResolvedManagedPriceFeed {
  /** The revision now serving, or null when none has ever validated. */
  blob: PriceFeedBlob | null;
  head: PriceFeedHead;
}

/** Stable, key-safe identity for a canonical feed URL. */
export function managedPriceUrlHash(url: string): string {
  return sliceSha256(url).slice(0, 32);
}

/**
 * The revision id for a fetched body: the server's strong ETag when it sends
 * one (sanitized to key-safe characters), else the body's own SHA-256.
 */
function revisionFor(etag: string | null, text: string): string {
  if (etag) {
    const safe = etag
      .replace(/^W\//u, "")
      .replace(/"/gu, "")
      .replace(/[^A-Za-z0-9._-]/gu, "");
    if (safe.length > 0 && safe.length <= 128) return safe;
  }
  return sliceSha256(text);
}

/** `BTC-USD`, `BTC-USD`, and `btc_usd` name the same instrument. */
const canonicalAlias = (alias: string): string =>
  alias.replace(/[^A-Za-z0-9]/gu, "").toUpperCase();

/**
 * A body that parses as prices can still be the wrong instrument: reject a
 * feed whose header alias disagrees with the URL (ignoring case and
 * separators), or whose commodity pair differs from the revision it would
 * replace (PRFAQ002 FAQ 6 and 14).
 */
function checkIdentity(
  validation: ManagedPriceValidation,
  alias: string,
  previous: PriceFeedBlob | null,
): ManagedPriceValidation {
  if (!validation.ok) return validation;
  const { feed } = validation;
  if (
    feed.alias !== null &&
    canonicalAlias(feed.alias) !== canonicalAlias(alias)
  ) {
    return {
      ok: false,
      reason: `feed alias ${feed.alias} does not match the requested ${alias}`,
      line: null,
    };
  }
  if (
    previous &&
    (previous.feed.commodity !== feed.commodity ||
      previous.feed.quote !== feed.quote)
  ) {
    return {
      ok: false,
      reason: `commodity pair changed from ${previous.feed.commodity}/${previous.feed.quote} to ${feed.commodity}/${feed.quote}`,
      line: null,
    };
  }
  return validation;
}

const EMPTY_HEAD: PriceFeedHead = {
  revision: null,
  nextRefreshAt: 0,
  lastError: null,
};

async function readBlob(
  cache: CacheHelper,
  urlHash: string,
  head: PriceFeedHead,
): Promise<PriceFeedBlob | null> {
  if (head.revision === null) return null;
  const blob = await cache.get<PriceFeedBlob>(
    CACHE_KEYS.ledger.priceFeedBlob(urlHash, head.revision),
  );
  return blob ?? null;
}

/**
 * Whether the cached state answers a load without a fetch: the refresh window
 * is still open, and either the blob is present or no revision has ever
 * validated (a head recording a failure inside its retry window). A head
 * whose blob was evicted falls through to a fetch instead of reporting nothing.
 */
function servable(
  head: PriceFeedHead,
  blob: PriceFeedBlob | null,
  now: number,
): boolean {
  return now < head.nextRefreshAt && (blob !== null || head.revision === null);
}

/**
 * Serve the feed at `url` from the cache, refreshing it when its window has
 * elapsed. Never throws for a feed problem: every failure is recorded on the
 * head and the caller receives whatever revision last validated.
 */
export async function resolveManagedPriceFeed(
  url: string,
  alias: string,
  deps: ManagedPriceFeedDeps,
): Promise<ResolvedManagedPriceFeed> {
  const { cache, config } = deps;
  const now = deps.now ?? Date.now;
  const urlHash = managedPriceUrlHash(url);
  const headKey = CACHE_KEYS.ledger.priceFeedHead(urlHash);

  const head = (await cache.get<PriceFeedHead>(headKey)) ?? EMPTY_HEAD;
  if (now() < head.nextRefreshAt) {
    const blob = await readBlob(cache, urlHash, head);
    if (servable(head, blob, now())) return { blob, head };
  }

  return lock.acquire(`${FEED_LOCK_PREFIX}${urlHash}`, async () => {
    // Another waiter may have refreshed while this one queued.
    const current = (await cache.get<PriceFeedHead>(headKey)) ?? EMPTY_HEAD;
    const previous = await readBlob(cache, urlHash, current);
    if (servable(current, previous, now())) {
      return { blob: previous, head: current };
    }

    const result = await fetchManagedPriceFeed(url, {
      etag: previous?.etag ?? null,
      timeoutMs: config.fetchTimeoutMs,
      maxBodyBytes: config.maxBodyBytes,
      fetchImpl: deps.fetchImpl,
    });
    const at = now();

    if (result.kind === "not-modified" && previous) {
      const refreshed: PriceFeedHead = {
        revision: previous.revision,
        nextRefreshAt: at + config.refreshMs,
        lastError: null,
      };
      await cache.set(headKey, refreshed, TTL.HOUR_24);
      log.debug("managed price feed unchanged", { url, revision: previous.revision });
      return { blob: previous, head: refreshed };
    }

    let message: string;
    if (result.kind === "fetched") {
      const validation = checkIdentity(
        validateManagedPriceText(result.text),
        alias,
        previous,
      );
      if (validation.ok) {
        const revision = revisionFor(result.etag, result.text);
        const blob: PriceFeedBlob = {
          url,
          revision,
          etag: result.etag,
          text: result.text,
          fetchedAt: at,
          feed: validation.feed,
        };
        const refreshed: PriceFeedHead = {
          revision,
          nextRefreshAt: at + config.refreshMs,
          lastError: null,
        };
        // The blob must exist before the head points at it; the superseded
        // blob can go while the head is written.
        await cache.set(
          CACHE_KEYS.ledger.priceFeedBlob(urlHash, revision),
          blob,
          TTL.HOUR_24,
        );
        await Promise.all([
          cache.set(headKey, refreshed, TTL.HOUR_24),
          previous && previous.revision !== revision
            ? cache.del(CACHE_KEYS.ledger.priceFeedBlob(urlHash, previous.revision))
            : Promise.resolve(),
        ]);
        log.info("managed price feed refreshed", {
          url,
          revision,
          prices: validation.feed.prices.length,
          latestObservedAt: validation.feed.latestObservedAt,
        });
        return { blob, head: refreshed };
      }
      message =
        validation.line === null
          ? `invalid feed: ${validation.reason}`
          : `invalid feed at line ${validation.line}: ${validation.reason}`;
    } else if (result.kind === "not-modified") {
      message = "server answered 304 without a cached revision to reuse";
    } else {
      message = `fetch failed (${result.reason}): ${result.message}`;
    }

    const degraded: PriceFeedHead = {
      revision: current.revision,
      nextRefreshAt: at + config.retryMs,
      lastError: message,
    };
    await cache.set(headKey, degraded, TTL.HOUR_24);
    log.warn("managed price feed refresh failed; serving the last revision", {
      url,
      previousRevision: previous?.revision ?? null,
      message,
    });
    return { blob: previous, head: degraded };
  });
}

File: cli/src/bea_engine/managed_price_cache.py
"""On-disk feed cache for managed price includes (w1/m29/t002).

Mirrors ADR 015 section 5 the way `managed_prices` mirrors sections 2–4: an
immutable blob per validated revision plus a small mutable head per URL, with
refresh driven by the head's time stamp rather than by key expiry. A failed
refresh records its error on the head while the previous revision keeps
serving, so a bad refresh never replaces a good cache entry with nothing.

Layout under `<cache>/bea/managed-prices/` (the same XDG base the write locks
use in `ledger/write.py`):

```
<sha256(url)[:32]>/
  head.json              # revision, next_refresh_at, last_error
  <revision>.beancount   # exact validated bytes
  <revision>.json        # etag, fetched_at, and the feed summary
```

The ETag lives on the blob, not the head, exactly as hosted: a head whose
blob went missing offers no stale ETag and simply fetches again. Writes are
atomic renames ordered blob-then-head, so concurrent CLI runs can only ever
expose a complete revision, never a half-written one; redundant fetches are
the only cost of two loads racing, and feeds are idempotent bytes.

Two modes beyond the hosted contract: offline resolution reads the head and
its blob only and never touches the network, and strict resolution fails the
load when a source is stale or unavailable instead of degrading quietly.
"""

from __future__ import annotations

import json
import os
import time
from dataclasses import asdict, dataclass
from datetime import datetime
from hashlib import sha256
from pathlib import Path
from typing import Literal
from urllib.request import OpenerDirector

from bea_engine.managed_prices import (
    FETCH_TIMEOUT_SECONDS,
    MAX_BODY_BYTES,
    FeedSummary,
    PricePoint,
    ValidFeed,
    check_feed_identity,
    feed_revision_id,
    fetch_managed_price_feed,
    validate_managed_price_text,
)
from bea_engine.protocol import LedgerError

REFRESH_SECONDS = 300
"""How long a validated revision serves before a conditional re-fetch."""

RETRY_SECONDS = 60
"""How long to wait after a failed refresh before trying again."""

STALE_SECONDS = 600
"""Observation age beyond which a source reads `stale` (ADR 015 section 8)."""


@dataclass(frozen=True)
class PriceFeedHead:
    """The mutable pointer: serving revision, next refresh, last error."""

    revision: str | None
    next_refresh_at: float
    last_error: str | None


@dataclass(frozen=True)
class PriceFeedBlob:
    """One immutable validated revision with the summary validation recorded."""

    url: str
    revision: str
    etag: str | None
    text: str
    fetched_at: float
    feed: FeedSummary


@dataclass(frozen=True)
class ResolvedFeed:
    """Whatever revision serves now, possibly none, with its head."""

    blob: PriceFeedBlob | None
    head: PriceFeedHead


def cache_root() -> Path:
    """Where feed revisions live — beside the write locks, under XDG cache."""
    base = Path(os.environ.get("XDG_CACHE_HOME") or Path.home() / ".cache").expanduser()
    return base / "bea" / "managed-prices"


def feed_dir(url: str, root: Path | None = None) -> Path:
    """The per-URL directory, keyed by a stable hash of the canonical URL."""
    return (root or cache_root()) / sha256(url.encode("utf-8")).hexdigest()[:32]


def _read_head(directory: Path) -> PriceFeedHead:
    try:
        raw = json.loads((directory / "head.json").read_text(encoding="utf-8"))
        return PriceFeedHead(
            revision=raw.get("revision"),
            next_refresh_at=float(raw.get("next_refresh_at", 0)),
            last_error=raw.get("last_error"),
        )
    except (OSError, ValueError, TypeError, AttributeError):
        return PriceFeedHead(revision=None, next_refresh_at=0.0, last_error=None)


def _read_blob(directory: Path, revision: str) -> PriceFeedBlob | None:
    try:
        text = (directory / f"{revision}.beancount").read_text(encoding="utf-8")
        raw = json.loads((directory / f"{revision}.json").read_text(encoding="utf-8"))
        feed = raw["feed"]
        return PriceFeedBlob(
            url=raw["url"],
            revision=raw["revision"],
            etag=raw.get("etag"),
            text=text,
            fetched_at=float(raw["fetched_at"]),
            feed=FeedSummary(
                alias=feed.get("alias"),
                commodity=feed["commodity"],
                quote=feed["quote"],
                source=feed.get("source"),
                revision=feed.get("revision"),
                latest_observed_at=feed.get("latest_observed_at"),
                prices=tuple(
                    PricePoint(
                        line=point["line"],
                        date=point["date"],
                        base=point["base"],
                        quote=point["quote"],
                        observed_at=point.get("observed_at"),
                    )
                    for point in feed.get("prices", [])
                ),
            ),
        )
    except (OSError, ValueError, TypeError, KeyError, AttributeError):
        return None


def _write_text(path: Path, text: str) -> None:
    """Write atomically, so a crash never leaves a half-written cache file."""
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_name(f".{path.name}.tmp")
    tmp.write_text(text, encoding="utf-8")
    os.replace(tmp, path)


def freshness(
    blob: PriceFeedBlob | None, now: float, stale_seconds: int = STALE_SECONDS
) -> Literal["recent", "stale", "unavailable"]:
    """Freshness computed at read time, never stored (ADR 015 section 8).

    Recent within the stale window of the latest `observed-at`, stale beyond
    it, unavailable when no revision validated. A feed with no parseable
    observed-at reads stale: without an observation age there is nothing
    fresh to claim, and the hosted side reads the same feed the same way.
    """
    if blob is None:
        return "unavailable"
    stamp = blob.feed.latest_observed_at
    if stamp is None:
        return "stale"
    moment = stamp[:-1] + "+00:00" if stamp.endswith("Z") else stamp
    try:
        age = now - datetime.fromisoformat(moment).timestamp()
    except ValueError:
        return "stale"
    return "recent" if age <= stale_seconds else "stale"


def resolve_feed(
    url: str,
    alias: str,
    *,
    root: Path | None = None,
    offline: bool = False,
    strict: bool = False,
    now: float | None = None,
    opener: OpenerDirector | None = None,
    timeout_seconds: int = FETCH_TIMEOUT_SECONDS,
    max_body_bytes: int = MAX_BODY_BYTES,
    refresh_seconds: int = REFRESH_SECONDS,
    retry_seconds: int = RETRY_SECONDS,
    stale_seconds: int = STALE_SECONDS,
) -> ResolvedFeed:
    """Serve the feed at `url` from the cache, refreshing past its window.

    Never fails for a feed problem in lenient mode: every failure is recorded
    on the head and the caller receives whatever revision last validated.
    Offline mode reads the head and its blob only and writes nothing. Strict
    mode raises naming the source when it is stale or unavailable.
    """
    at = time.time() if now is None else now
    directory = feed_dir(url, root)
    head = _read_head(directory)
    previous = _read_blob(directory, head.revision) if head.revision is not None else None
    if offline:
        if previous is None and head.revision is not None:
            head = PriceFeedHead(revision=None, next_refresh_at=0.0, last_error=head.last_error)
        resolved = ResolvedFeed(blob=previous, head=head)
        _enforce_strict(url, resolved, at, stale_seconds, strict)
        return resolved
    if at < head.next_refresh_at and (previous is not None or head.revision is None):
        resolved = ResolvedFeed(blob=previous, head=head)
        _enforce_strict(url, resolved, at, stale_seconds, strict)
        return resolved

    result = fetch_managed_price_feed(
        url,
        etag=previous.etag if previous else None,
        opener=opener,
        timeout_seconds=timeout_seconds,
        max_body_bytes=max_body_bytes,
    )
    if result.kind == "not-modified" and previous is not None:
        refreshed = PriceFeedHead(revision=previous.revision, next_refresh_at=at + refresh_seconds, last_error=None)
        _write_text(directory / "head.json", json.dumps(asdict(refreshed)))
        resolved = ResolvedFeed(blob=previous, head=refreshed)
        _enforce_strict(url, resolved, at, stale_seconds, strict)
        return resolved

    message: str | None = None
    if result.kind == "fetched":
        pair = (previous.feed.commodity, previous.feed.quote) if previous else None
        validation = check_feed_identity(validate_managed_price_text(result.text), alias, pair)
        if validation.ok:
            assert isinstance(validation, ValidFeed)
            revision = feed_revision_id(result.etag, result.text)
            blob = PriceFeedBlob(
                url=url,
                revision=revision,
                etag=result.etag,
                text=result.text,
                fetched_at=at,
                feed=validation.feed,
            )
            # The blob must exist before the head points at it; the superseded
            # blob can go while the head is written.
            _write_text(directory / f"{revision}.beancount", result.text)
            _write_text(
                directory / f"{revision}.json",
                json.dumps(
                    {
                        "url": url,
                        "revision": revision,
                        "etag": result.etag,
                        "fetched_at": at,
                        "feed": {
                            **asdict(validation.feed),
                            "prices": [asdict(point) for point in validation.feed.prices],
                        },
                    }
                ),
            )
            refreshed = PriceFeedHead(revision=revision, next_refresh_at=at + refresh_seconds, last_error=None)
            _write_text(directory / "head.json", json.dumps(asdict(refreshed)))
            if previous is not None and previous.revision != revision:
                (directory / f"{previous.revision}.beancount").unlink(missing_ok=True)
                (directory / f"{previous.revision}.json").unlink(missing_ok=True)
                # Per-ledger effective texts key off the revision; they die with it.
                for effective in directory.glob(f"{previous.revision}.effective.*.beancount"):
                    effective.unlink(missing_ok=True)
            resolved = ResolvedFeed(blob=blob, head=refreshed)
            _enforce_strict(url, resolved, at, stale_seconds, strict)
            return resolved
        message = (
            f"invalid feed: {validation.reason}"
            if validation.line is None
            else f"invalid feed at line {validation.line}: {validation.reason}"
        )
    elif result.kind == "not-modified":
        message = "server answered 304 without a cached revision to reuse"
    else:
        message = f"fetch failed ({result.reason}): {result.message}"

    degraded = PriceFeedHead(revision=head.revision, next_refresh_at=at + retry_seconds, last_error=message)
    _write_text(directory / "head.json", json.dumps(asdict(degraded)))
    resolved = ResolvedFeed(blob=previous, head=degraded)
    _enforce_strict(url, resolved, at, stale_seconds, strict)
    return resolved


def _enforce_strict(url: str, resolved: ResolvedFeed, at: float, stale_seconds: int, strict: bool) -> None:
    """Fail the load, naming the source, when strict mode forbids degrading."""
    if not strict:
        return
    if resolved.blob is None:
        cause = resolved.head.last_error or "no cached revision"
        raise LedgerError(f"managed price source {url} is unavailable in strict mode: {cause}.")
    if freshness(resolved.blob, at, stale_seconds) == "stale":
        stamp = resolved.blob.feed.latest_observed_at or "unknown observation time"
        raise LedgerError(
            f"managed price source {url} is stale in strict mode: latest observation {stamp} "
            f"is older than {stale_seconds} seconds."
        )


def managed_source_for_path(path: Path, root: Path | None = None) -> str | None:
    """The feed URL a cache path belongs to, or None outside the feed cache.

    The read-only write boundary uses this to name the managed source a
    write tried to target.
    """
    try:
        path.resolve().relative_to((root or cache_root()).resolve())
    except (OSError, ValueError):
        return None
    for sidecar in sorted(path.parent.glob("*.json")):
        if sidecar.name == "head.json":
            continue
        try:
            url = json.loads(sidecar.read_text(encoding="utf-8")).get("url")
        except (OSError, ValueError, AttributeError):
            continue
        if isinstance(url, str):
            return url
    return None


def zero_next_refresh(url: str, root: Path | None = None) -> PriceFeedHead:
    """Force the next load to re-resolve: the manual refresh of ADR 015 section 5."""
    directory = feed_dir(url, root)
    head = _read_head(directory)
    refreshed = PriceFeedHead(revision=head.revision, next_refresh_at=0.0, last_error=head.last_error)
    _write_text(directory / "head.json", json.dumps(asdict(refreshed)))
    return refreshed
